### PR TITLE
In-app inbox extensions (updated hook, new fields, summary method)

### PR DESCRIPTION
**Pull request by ftcvlad**
### Description of Changes

Port latest updates from Swift. This PR matches https://github.com/Kumulos/KumulosSdkSwift/pull/73, see description there.

### Breaking Changes

-   None

### Release Checklist

Prepare:

-   [x] Detail any breaking changes. Breaking changes require a new major version number
-   [x] Check `pod lib lint` passes
-   [x] Check all targets (iOS, extension, macOS, statics) build
-   [x] Install branch via Cocoapods into empty project & verify can import SDK (only needed if adding/removing files)

Bump versions in:

-   [x] `Sources/Kumulos+Stats.m`
-   [x] `KumulosSdkObjectiveC.podspec`
-   [x] `KumulosSdkObjectiveCExtension.podspec`
-   [x] All relevant build targets
-   [x] `README.md`

Release:

-   [ ] Squash and merge to master
-   [ ] Delete branch once merged
-   [ ] ~~Create tag from master matching chosen version~~ => hold
-   [ ] ~~Run `pod trunk push` to publish to CocoaPods~~ => hold

### CODE REVIEW

File: Sources/InApp/KSInAppHelper.m [hunk 1]
```diff
@@ -37,6 +37,7 @@ @interface KSInAppHelper ()
 @implementation KSInAppHelper
 
 int const STORED_IN_APP_LIMIT = 50;
+InboxUpdatedHandlerBlock _inboxUpdatedHandlerBlock = nil;
```
Comment (cgwyllie) [inline Sources/InApp/KSInAppHelper.m:40]: For these global C vars, should they be static for retention?

Can't remember how we've approached this before. Believe some of the SDKs like Unity perhaps has some global closures to capture swizzled push delegates into.

Alt perhaps a static (class) property on KumulosInApp?

File: Sources/InApp/KSInAppHelper.m [hunk 2]
```diff
@@ -485,13 +504,20 @@ - (NSMutableArray*) evictMessagesExceedingLimit:(NSManagedObjectContext* _Nonnul
     return messages;
 }
 
-- (void)handleMessageOpened:(KSInAppMessage *)message {
-    [self markInboxItemRead:message.id shouldWait:false];
-    
+- (void)handleMessageOpened:(KSInAppMessage*)message {
+    BOOL markedRead = NO;
+    if (message.readAt == nil){
+        markedRead =  [self markInboxItemRead:message.id shouldWait:false];
```
Comment (cgwyllie) [inline Sources/InApp/KSInAppHelper.m:510]: `shouldWait: false` would imply markedRead never has a useful value?

File: Sources/InApp/KSInAppHelper.m [hunk 3]
```diff
@@ -698,30 +726,97 @@ - (BOOL)markInboxItemRead:(NSNumber*)withId shouldWait:(BOOL)shouldWait{
...
+    if (self.messagesContext == nil){
+        [self fireInboxSummaryCallback:inboxSummaryBlock summary:nil];
+        return;
+    }
+
+    [self.messagesContext performBlock:^{
+        NSManagedObjectContext* context = self.messagesContext;
+
+        NSFetchRequest* request = [NSFetchRequest fetchRequestWithEntityName:@"Message"];
+        [request setIncludesPendingChanges:YES];
```
Comment (cgwyllie) [inline Sources/InApp/KSInAppHelper.m:785]: Should this be `NO`? 

https://github.com/Kumulos/KumulosSdkSwift/pull/73/files#diff-396cde03f399b0e07b35bf09fca1e16b42fb7416969837690fb32b852287d881R824

File: Sources/InApp/KSInAppHelper.m (hunk 2 above)
Comment (ftcvlad) [reply Sources/InApp/KSInAppHelper.m:510]: if we dont wait we assume markingRead was successful and return YES immediately. could assign on 2 lines, but same

File: Sources/InApp/KSInAppHelper.m (hunk 3 above)
Comment (ftcvlad) [reply Sources/InApp/KSInAppHelper.m]: o_o oops

File: Sources/InApp/KSInAppHelper.m (hunk 1 above)
Comment (ftcvlad) [reply Sources/InApp/KSInAppHelper.m:40]: There was a problem with having it on KumulosInApp -- _inboxUpdatedHandlerBlock has to be accessible from our SDK, but not accessible from the outside world. how do you achieve protected in objc without e.g. having another header file?

inAppHelper is stored as a static property on Kumulos. _inboxUpdatedHandlerBlock is an instance variable, strong by default (from [there](https://stackoverflow.com/questions/2255861/property-retain-assign-copy-nonatomic-in-objective-c) : `By default all instance variables and local variables are strong pointers.`)

could make it static in KSInAppHelper as well

Comment (cgwyllie) [reply Sources/InApp/KSInAppHelper.m:40]: Ah, sorry my bad, I didn't clock it was a property in the `@implementation`, all good here then.

Comment (cgwyllie) [reply Sources/InApp/KSInAppHelper.m:40]: (and yes, in ObjC no real such thing as 'protected' except headers that aren't public, so we kind of do protected fields with an 'anonymous' category)

File: Sources/InApp/KSInAppHelper.m (hunk 2 above)
Comment (cgwyllie) [reply Sources/InApp/KSInAppHelper.m:510]: OK, ah so it's more about whether readAt was nil, so should we run the handler. Seems fine then.

File: Sources/InApp/KSInAppHelper.m [hunk 4]
```diff
@@ -698,30 +726,97 @@ - (BOOL)markInboxItemRead:(NSNumber*)withId shouldWait:(BOOL)shouldWait{
...
+}
+
+- (void)maybeRunInboxUpdatedHandler:(BOOL)inboxNeedsUpdate {
+    if (!inboxNeedsUpdate){
+        return;
+    }
+
+    if (_inboxUpdatedHandlerBlock != nil){
+        dispatch_async(dispatch_get_main_queue(), _inboxUpdatedHandlerBlock);
+    }
```
Comment (cgwyllie) [inline Sources/InApp/KSInAppHelper.m:772]: Is there any possibility by the time the main queue runs the handler, it's been set to nil?

Probably a super rare edge case (like maybe unloading thing screen setting handler nil on unload whilst handler runs code), but not clear if the block will be retained with a reference or if we're passing a pointer here?

(also presume set/read isn't atomic, properties have some attribute (atomic) which can make access to them synchronised, but also not sure that's enough to help)

Perhaps pass anon block to main queue with the guard & invocation inside it?

File: Sources/InApp/KSInAppHelper.m (hunk 2 above)
Comment (ftcvlad) [reply Sources/InApp/KSInAppHelper.m:510]: yes, it's about readAt

File: Sources/InApp/KSInAppHelper.m (hunk 4 above)
Comment (ftcvlad) [reply Sources/InApp/KSInAppHelper.m:772]: [here](https://developer.apple.com/documentation/dispatch/1453057-dispatch_async) it says  
`The block to submit to the target dispatch queue. This function performs Block_copy and Block_release on behalf of callers` -- so looks the block is retained?

i think atomic is default

File: Sources/KumulosInApp.m [hunk 5]
```diff
@@ -36,7 +43,17 @@ - (BOOL) isRead {
     return self.readAt != nil;
 }
 
+@end
+
+@implementation InAppInboxSummary
+
++ (instancetype) init:(int)totalCount unreadCount:(int)unreadCount {
+    InAppInboxSummary* item = [InAppInboxSummary new];
```
Comment (cgwyllie) [inline Sources/KumulosInApp.m:51]: Should this initializer be calling super init? Stypical initializer looks like:

```
if ([super init]) {
  // Do work
}
```

File: Sources/KumulosInApp.m [hunk 6]
```diff
@@ -18,7 +18,14 @@ + (instancetype) fromInboxItemEntity:(KSInAppMessageEntity*)entity {
     item->_availableTo = [entity.inboxTo copy];
     item->_dismissedAt = [entity.dismissedAt copy];
     item->_readAt = [entity.readAt copy];
-    
+    item->_data= [entity.data copy];
+    if (entity.sentAt != nil){
+        item->_sentAt = [entity.sentAt copy];
+    }
+    else{
+        item->_sentAt = [entity.updatedAt copy];
```
Comment (cgwyllie) [inline Sources/KumulosInApp.m:26]: ☕ 🍝

Comment (cgwyllie) [reply Sources/KumulosInApp.m:26]: Ah no, nvm.

File: Sources/KumulosInApp.m (hunk 5 above)
Comment (cgwyllie) [reply Sources/KumulosInApp.m:51]: Ah ok, maybe nvm sorry see this is a class method.

**Pull request merged**